runtime/jcl/common/thread.cpp

```diff
@@ -519,6 +519,105 @@ Java_java_lang_Thread_registerNatives(JNIEnv *env, jclass clazz)
 	clearNonZAAPEligibleBit(env, clazz, natives, numNatives);
 #endif /* J9VM_OPT_JAVA_OFFLOAD_SUPPORT */
 }
+
+/* private native void notifyJvmtiMountBegin(boolean firstMount); */
+void JNICALL
+Java_java_lang_VirtualThread_notifyJvmtiMountBegin(JNIEnv *env, jobject thread, jboolean firstMount)
+{
+	J9VMThread *currentThread = (J9VMThread *)env;
+	J9JavaVM *vm = currentThread->javaVM;
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+	J9MemoryManagerFunctions *mmFuncs = vm->memoryManagerFunctions;
+
+	if (firstMount) {
+		vmFuncs->internalEnterVMFromJNI(currentThread);
+
+		omrthread_monitor_enter(vm->liveVirtualThreadListMutex);
+		if (NULL == vm->liveVirtualThreadList) {
+			J9Class *virtualThreadClass = J9OBJECT_CLAZZ(currentThread, J9_JNI_UNWRAP_REFERENCE(thread));
+			j9object_t rootVirtualThread = NULL;
+
+			/* Allocate empty virtual thread and create a global reference to it as root for the linked list.
+			 * This prevents the root reference from becoming stale if the GC moves the object.
+			 */
+			rootVirtualThread = mmFuncs->J9AllocateObject(currentThread, virtualThreadClass, J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE);
+			if (NULL != rootVirtualThread) {
+				jobject globalRef = vmFuncs->j9jni_createGlobalRef((JNIEnv *)currentThread, rootVirtualThread, JNI_FALSE);
+				if (NULL != globalRef) {
+					vm->liveVirtualThreadList = (j9object_t *)globalRef;
+
+					/* Set linkNext/linkPrevious to itself. */
+					J9OBJECT_OBJECT_STORE(currentThread, rootVirtualThread, vm->virtualThreadLinkNextOffset, rootVirtualThread);
+					J9OBJECT_OBJECT_STORE(currentThread, rootVirtualThread, vm->virtualThreadLinkPreviousOffset, rootVirtualThread);
+				} else {
+					vmFuncs->setNativeOutOfMemoryError(currentThread, 0, 0);
+				}
+			} else {
+				vmFuncs->setHeapOutOfMemoryError(currentThread);
+			}
+		}
+
+		if (NULL != vm->liveVirtualThreadList) {
```

Comment:
We want to execute both blocks if the global ref has not been created yet: to create it, then add the virtual thread. Also if something fails when creating the root, we want to skip adding the virtual thread.

Reply:
`vm->liveVirtualThreadList` is only set when the global ref and root are successfully created. Otherwise, it will stay `NULL` and an error will be thrown. It won't be possible to reach the `else` block unless `vm->liveVirtualThreadList` is correctly set.

Reply:
If we convert this to else then the first virtual thread will add the root but not itself to the list.